core/common/src/grpc_interceptor.rs

```diff
@@ -0,0 +1,238 @@
+
+    /// Implementation of tower's Service trait's poll_ready method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    /// Implementation of tower's Service trait's call method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn call(
+        &mut self,
+        mut request: http::request::Request<tonic::transport::Body>,
+    ) -> Self::Future {
+        let interceptor = (self.interceptor_factory)();
+
+        let (service_name, method_name) = Self::retrieve_grpc_names_from_uri(request.uri());
+        let is_applicable = interceptor.is_applicable(&service_name, &method_name);
+
+        if is_applicable && interceptor.must_handle_request() {
+            let (parts, body) = request.into_parts();
+            let mut body_bytes: Bytes =
+                match futures::executor::block_on(hyper::body::to_bytes(body)) {
+                    Ok(bytes) => bytes,
+                    Err(err) => {
+                        return Box::pin(async move {
+                            Err(Box::new(err) as Box<dyn std::error::Error + Sync + Send>)
+                        })
+                    }
+                };
+            let protobuf_message_bytes: Bytes = body_bytes.split_off(GRPC_HEADER_LENGTH);
+            let grpc_header_bytes = body_bytes;
+            let new_protobuf_message_bytes: Bytes = match interceptor.handle_request(
+                &service_name,
+                &method_name,
+                protobuf_message_bytes,
+            ) {
+                Ok(bytes) => bytes,
+                Err(err) => return Box::pin(async move { Err(err) }),
+            };
+            let new_body_chunks: Vec<Result<_, std::io::Error>> =
+                vec![Ok(grpc_header_bytes), Ok(new_protobuf_message_bytes)];
+            let stream = futures_util::stream::iter(new_body_chunks);
+            let new_body = tonic::transport::Body::wrap_stream(stream);
+            request = http::request::Request::from_parts(parts, new_body);
+        }
+
+        let fut = self.service.call(request);
+
+        Box::pin(async move {
+            match fut.await {
+                Ok(response) => {
+                    if is_applicable && interceptor.must_handle_response() {
+                        let (parts, body) = response.into_parts();
+                        let mut body_bytes = match hyper::body::to_bytes(body).await {
+                            Ok(bytes) => bytes,
+                            Err(err) => {
+                                return Err(
+                                    Box::new(err) as Box<dyn std::error::Error + Sync + Send>
+                                )
+                            }
+                        };
+                        let protobuf_message_bytes = body_bytes.split_off(GRPC_HEADER_LENGTH);
```

Comment:
much of the logic for messing with the GRPC data here is very similar to above and it's weird enough that I think it should be encapsulated. I'd suggest writing some common functions, something like these (note they might be incomplete since I'm coding in a review comment, but they should be close):

```rust
async fn split_grpc_bytes<T: HttpBody>(body: T) -> (Bytes, Bytes) {
    let mut body_bytes = ...; // Note I've suggested a change to the logic for this in another comment

    // My personal preference is to use split_to and extract the header out, but either way works fine
    let protobuf_header = body_bytes.split_to(GRPC_HEADER_LENGTH);
    (protobuf_header, body_bytes)
}

fn merge_grpc_bytes(header: Bytes, body: Bytes) -> tonic::transport::Body {
    let new_body_chunks: Vec<Result<_, std::io::Error>> =
        vec![Ok(header), Ok(body)];
    let stream = futures_util::stream::iter(new_body_chunks);
    tonic::transport::Body::wrap_stream(stream)
}
```

on the surface there's more that looks like it could be made common but I think that's just because the request and response objects have a similar API, so it would be rough to try to make much more of the code common

Reply:
There will be other PRs to make improvements like this.  It's important to get this PR in today so that Devin can start using this.